cfep-20.md

```diff
@@ -0,0 +1,97 @@
+ - This CFEP will be much easier to execute upon if we improve the conda tooling around the support for multiple outputs for recipes.
+
+## Motivation
+
+At the moment, there is not a common pattern to split packages in multiple output including C/C++ headers, JS source maps, or artifacts of build systems that are not required at runtime but are necessary at build time. Packages have been split in various ways, but adopting a convention will allow us to create tooling, idioms, and best practices on how to deal with such split packages.
+
+## Output names and what goes where
+
+The proposal is to split most packages in several outputs. For example, package `foobar` may be split into
+
+ - `foobar`
+ - `foobar-dev`
+ - `foobar-dbg`
+ - `foobar-doc`
+
+### `foobar`, the main runtime package
+
+The `foobar` package should only contain the parts of the package that are needed at runtime, such as 
+
+ - Shared objects
+ - Executables
+ - Other assets needed at runtime to use the package
+
+It may not include headers, static libraries, debug symbols, JavaScript source maps, or artifacts from the build systems such as pkg-config or cmake. These exclusions may be checked in CI.
+
+### `foobar-dev`, the development package
+
+The `foobar-dev` should contain headers, build system artifacts, and possible static libraries.
+
+It may not include shared objects or executables, or any other content required at runtime by the `foobar` package.
+Also, debug symbols and source maps should be in the `foobar-dbg` package.
+
+### `foobar-dbg`, the debug package
+
+The `foobar-dbg` should contain the debug symbols, or other similar artifacts, such as JS sourcemaps, which are not required at runtime or for building on top of the package, but are needed to troubleshoot issues with the package at runtime.
+
+It may not include any of the content of the `foobar` or `foobar-dev` packages. The `foobar-dbg` package is really a companion package to `foobar`, and should not e.g. include a debug build of the same package.
+
+### `foobar-doc`, the documentation package
+
+The `foobar-doc` package contains built documentation for `foobar`.
+
+### Special cases
+
+Some packages may only have a `-dev` package and no runtime package, such as header-only libraries, or libraries that do not produce a shared object.
+
+## Dependencies and exports
+
+### Dependencies between `foobar`, `foobar-dev`, and `foobar-dbg`
+
+ - We expect `foobar-dbg` to always list `foobar` as a run dependency, pinned to the same version.
+ - However, `foobar-doc` does not require `foobar`.
+ - Also, `foobar-dev` will typically depend on `foobar` (if it exists), pinned to the same version.
+
+Note: while we may expect usecases where for e.g. statically linking with libfoobar, we only need foobar-dev, it is probably a good pattern to have foobar-dev always depend on foobar to not break expectations.
```

Comment:
This is another reason to split the static library into a separate package from the headers

Reply:
I don't understand this argument. The static libraries would be packaged in the `-dev` package which would never be a dependency of non-dev packages.

Reply:
How would we make sure that static packages are never used if we have `-dev` packages in host when building? If the static packages are also brought in by `-dev` we have no idea whether the static package was used or not

Reply:
If we are going to change how static libraries are handled, we need to update [CFEP 18]( https://github.com/conda-forge/cfep/blob/84fc0ceae274d155639ef1561e54fd57b2b91335/cfep-18.md ) as well since that already prescribes how static libraries should be handled.

Reply:
> If the static packages are also brought in by -dev we have no idea whether the static package was used or not.

We don't know by just looking at the recipe meta.yaml. The recipe author should know what libraries they are linking with hopefully.

Reply:
At the moment, we know because `-static` are separate packages and if they are not in the `host` requirements (directly or indirectly) they are not used. By adding it to `-dev` it will always be in `host` indirectly and we will not know. Why change the current situation?

Reply:
> Why change the current situation?

The rationale for changing is that static libraries and headers have roughly the same implications with respect to binary compatibility, and we will end up applying the same pattern for both with respect to exports (`run exports`, and eventually `run_constraints exports`). Also, it does not make sense to have the static libraries without the headers.

Reply:
I'm not talking about binary compatibility. What I mean is that having the static libraries in `host` might make a downstream link to the static library instead of the shared library which we do not want. We always want shared libraries to be used.

> Also, it does not make sense to have the static libraries without the headers.

It makes sense to have headers without static libraries. `-static` output can have a dependency on `-dev`.

Reply:
> What I mean is that having the static libraries in host might make a downstream link to the static library instead of the shared library which we do not want. We always want shared libraries to be used.

You fear that static libraries may be linked against ***by accident***?